General validator function for checking dimensions and coordinates

niksirbi opened the pull request:
## Description

**What is this PR**

- [ ] Bug fix
- [x] Addition of a new feature
- [ ] Other

**Why is this PR needed?**
See #292 and #291 

**What does this PR do?**
Turns a previously private function for validating the existence of dimensions and coordinates into a public function in `validators/arrays.py/validate_dims_coords`. This function is then used in both `kinematics.py` and `vector.py`.
In kinematics, it's now also used to assert cartesian coordinates for computing displacement, velocity and acceleration.

## How has this PR been tested?
Existing test were already indirectly covering this functionality, but I decided to also add some unit tests that directly test this new validator. These tests are also much stricter in checking that the thrown exception has the right error message.

## Is this a breaking change?

Nope.

## Does this PR require an update to the documentation?

No, API docs will be automatically updated.

## Checklist:

- [x] The code has been tested locally
- [x] Tests have been added to cover all new functionality
- [n/a] The documentation has been updated to reflect any changes
- [x] The code has been formatted with [pre-commit](https://pre-commit.com/)


## Update 2024-09-16
This PR also exposes `compute_time_derivative` as a public function, and thus closes #311, following discussions in #291. This function is not restricted to Cartesian space.

lochhh reviewed (commented):
Thanks @niksirbi ! This is a great start to the much needed validators module. The updated suggestions are minor - happy for you to take or leave.
~~While this PR aims to address #291 as well, I have a different opinion on this and have detailed it [there](https://github.com/neuroinformatics-unit/movement/issues/291#issuecomment-2344547483).~~

lochhh commented on movement/analysis/kinematics.py:
```diff
@@ -42,7 +43,7 @@ def compute_displacement(data: xr.DataArray) -> xr.DataArray:
 
     """
-    _validate_time_dimension(data)
+    validate_dims_coords(data, {"time": [], "space": ["x", "y"]})
```
This will allow for 1-,2-,3-D in cartesian

niksirbi commented:
Thanks for the review @lochhh!

I've addressed your comments, and I've additionally exposes `compute_time_derivative` as a public function, as discussed in #291. So this PR now additionally closes #311.

Re-requesting review to double-check that you're happy with the way the new public function works.

lochhh commented:
> Thanks for the review @lochhh!
> 
> I've addressed your comments, and I've additionally exposed `compute_time_derivative` as a public function, as discussed in #291. So this PR now additionally closes #311.
> 
> Re-requesting review to double-check that you're happy with the way the new public function works.

Thanks again @niksirbi . Made a couple of small changes: refactor test + update old refs to `xr.differentiate` in `compute_velocity` and `compute_acceleration` to point directly to the newly exposed  `compute_time_derivative`.

niksirbi commented:
Thanks @lochhh! Nice final touches.
